File: app/schemas/scanner.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict

# ── Watchlist ──


class WatchlistItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    tradingsymbol: str
    exchange: str
    notes: str
    added_at: datetime


class WatchlistItemCreateRequest(BaseModel):
    tradingsymbol: str
    exchange: str = "NSE"
    notes: str = ""


# ── Signals ──


class SignalResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    tradingsymbol: str
    exchange: str
    signal_type: str
    timeframe: str
    entry_price: float
    stop_loss: float
    target_price: float
    confidence: float
    indicators: dict[str, Any]
    rationale: str
    status: str
    created_at: datetime
    expired_at: datetime | None
    outcome: str | None = None
    actual_exit_price: float | None = None
    actual_rr: float | None = None
    outcome_at: datetime | None = None


class SignalUpdateRequest(BaseModel):
    status: str  # executed / expired


class ExpireAllRequest(BaseModel):
    tradingsymbols: list[str] | None = None  # if provided, only expire these


class ExpireAllResponse(BaseModel):
    expired: int


# ── Scanner ──


class SymbolInput(BaseModel):
    tradingsymbol: str
    exchange: str = "NSE"


class ScanRequest(BaseModel):
    timeframe: str = "15minute"
    symbols: list[SymbolInput] | None = None


class ScanResultItem(BaseModel):
    tradingsymbol: str
    exchange: str
    signal_type: str
    entry_price: float
    stop_loss: float
    target_price: float
    confidence: float
    rationale: str


class ScanResponse(BaseModel):
    scanned: int
    signals_generated: int
    results: list[ScanResultItem]
    errors: list[str] = []


class ScanStatusResponse(BaseModel):
    last_scan: str | None = None
    status: str | None = None
    symbols_scanned: int = 0
    signals_generated: int = 0
    errors_count: int = 0
    duration_seconds: float = 0.0

File: frontend/src/types/scanner.ts
export interface WatchlistItem {
  id: number;
  tradingsymbol: string;
  exchange: string;
  notes: string;
  added_at: string;
}

export interface WatchlistItemCreateRequest {
  tradingsymbol: string;
  exchange?: string;
  notes?: string;
}

export interface Signal {
  id: number;
  tradingsymbol: string;
  exchange: string;
  signal_type: string;
  timeframe: string;
  entry_price: number;
  stop_loss: number;
  target_price: number;
  confidence: number;
  indicators: Record<string, unknown>;
  rationale: string;
  status: string;
  created_at: string;
  expired_at: string | null;
}

export interface SignalUpdateRequest {
  status: string;
}

export interface SymbolInput {
  tradingsymbol: string;
  exchange: string;
}

export interface ScanRequest {
  timeframe?: string;
  symbols?: SymbolInput[];
}

export interface ScanResultItem {
  tradingsymbol: string;
  exchange: string;
  signal_type: string;
  entry_price: number;
  stop_loss: number;
  target_price: number;
  confidence: number;
  rationale: string;
}

export interface ScanResponse {
  scanned: number;
  signals_generated: number;
  results: ScanResultItem[];
  errors: string[];
}

export interface InstrumentSearchResult {
  tradingsymbol: string;
  name: string;
  exchange: string;
}

export interface ScanStatus {
  last_scan: string | null;
  status: string | null;
  symbols_scanned: number;
  signals_generated: number;
  errors_count: number;
  duration_seconds: number;
}
